hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/SimpleRegionNormalizer.java

```diff
@@ -71,38 +76,81 @@
   static final String MERGE_MIN_REGION_SIZE_MB_KEY = "hbase.normalizer.merge.min_region_size.mb";
   static final int DEFAULT_MERGE_MIN_REGION_SIZE_MB = 1;
 
-  private Configuration conf;
   private MasterServices masterServices;
-  private boolean splitEnabled;
-  private boolean mergeEnabled;
-  private int minRegionCount;
-  private Period mergeMinRegionAge;
-  private int mergeMinRegionSizeMb;
+
+  /** Ensure configuration changes are applied atomically. */
+  private final ReadWriteLock configUpdateLock = new ReentrantReadWriteLock();
+  @GuardedBy("configUpdateLock") private Configuration conf;
```

Comment:
This is cool. I believe this will help us avoid partial config update!

Reply:
Indeed ! Wondering if we can ensure all implementors of `ConfigurationObserver` can start using such lock for atomic updates of non-final fields (of course not as part of this Jira :) )

Reply:
I'm not sure about "all" uses, it just seemed prudent for this use. I'm also not sure our static analysis tools honor this `GuardedBy` annotation, I have a TODO for myself to track this down and see if it's really supported.

Reply:
I have learnt about `GuardedBy` for the first time on this PR, so I am also not sure if static analysis tools really require some work.

Reply:
It seems SpotBugs does have a bug description for the improper use of `GuardedBy`, https://spotbugs.readthedocs.io/en/latest/bugDescriptions.html#is-field-not-guarded-against-concurrent-access-is-field-not-guarded. I wonder if it's actually implemented ;)